Support building on macOS

doinkythederp opened the pull request:
This pull request refactors the build script to work properly on macOS.

## The problem

The current Cake build script doesn't work on ARM Macs because the Cake Git APIs are only complied for x86/x86_64, not arm64. There are also some places that incorrectly handle differences between macOS and Linux.

## The solution

[Cake Frosting](https://cakebuild.net/docs/running-builds/runners/cake-frosting) is a newer option for setting up a Cake project which drops the need for a `build.cake` DSL and just gives you a normal `csproj` to write your build logic in. It has the same user-facing API as Cake without the aforementioned issues when building on ARM. It also has better IntelliSense in multi-platform IDEs like JetBrains Rider.

I re-organized the build script to use the Cake Frosting style and improved the logic for finding tool paths on macOS.

I also added a JetBrains Rider config which improves the development experience on macOS because it allows developers to open the Visual Studio solution without being on Windows.

## Structural changes

I moved the free-standing functions like `GetVersion`, etc. into the new `BuildContext` class so they could be accessed from any task. I also refactored the `MakeIn` function into a `MakeTask` class which is more ergonomic when using Cake Frosting. I also put many of the top-level variables into a `BuildPaths` class.

## TODO

- [x] Make it build on macOS
  - [x] Confirm that it actually starts up, works, etc
- [x] Update CI
- [x] Confirm that it still builds (and works) on Windows and Linux.
- [x] Make the tests pass
  - ~~Currently they aren't passing because I have Dotnet 9.0 installed while CKAN targets Dotnet 8.0, so they're refusing to run.~~

doinkythederp commented:
This is ready for review, I'd love it if someone could confirm this doesn't break builds for them!

doinkythederp commented:
I've confirmed that this doesn't affect building on Linux on account of the CI still working:

https://github.com/doinkythederp/CKAN/actions/runs/13471224444

HebaruSan commented:
After fixing that, it says "Could not resolve MSBuild":

```

----------------------------------------
Setup
----------------------------------------

========================================
Restore
========================================
  Determining projects to restore...
  Restored /media/sf_CKAN/Tests/Tests.csproj (in 2.95 min).
  Restored /media/sf_CKAN/Netkan/CKAN-netkan.csproj (in 1.95 sec).
  Restored /media/sf_CKAN/GUI/CKAN-GUI.csproj (in 5.88 sec).
  Restored /media/sf_CKAN/Core/CKAN-core.csproj (in 8.32 sec).
  Restored /media/sf_CKAN/ConsoleUI/CKAN-ConsoleUI.csproj (in 161 ms).
  Restored /media/sf_CKAN/Cmdline/CKAN-cmdline.csproj (in 4.44 sec).
  Restored /media/sf_CKAN/AutoUpdate/CKAN-autoupdate.csproj (in 100 ms).

========================================
Generate-GlobalAssemblyVersionInfo
========================================

========================================
Build
========================================
MSBuild version 17.8.5+b5265ef37 for .NET
  Determining projects to restore...
  Restored /media/sf_CKAN/Core/CKAN-core.csproj (in 6.58 sec).
  CKAN-core -> /media/sf_CKAN/_build/out/CKAN/Debug/bin/netstandard2.0/CKAN.dll
  CKAN-core -> /media/sf_CKAN/_build/out/CKAN/Debug/bin/net48/CKAN.dll
/media/sf_CKAN/Core/Repositories/RepositoryData.cs(159,29): warning SYSLIB0050: 'StreamingContextStates' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
/media/sf_CKAN/Core/Repositories/RepositoryData.cs(158,35): warning SYSLIB0050: 'StreamingContext.StreamingContext(StreamingContextStates, object?)' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
/media/sf_CKAN/Core/Registry/RegistryManager.cs(316,48): warning SYSLIB0050: 'StreamingContextStates' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
/media/sf_CKAN/Core/Registry/RegistryManager.cs(316,27): warning SYSLIB0050: 'StreamingContext.StreamingContext(StreamingContextStates, object?)' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
  CKAN-core -> /media/sf_CKAN/_build/out/CKAN/Debug/bin/net8.0/CKAN.dll
  Successfully created package '/media/sf_CKAN/_build/out/CKAN/Debug/bin/CKAN.1.35.3.nupkg'.

Build succeeded.

/media/sf_CKAN/Core/Repositories/RepositoryData.cs(159,29): warning SYSLIB0050: 'StreamingContextStates' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
/media/sf_CKAN/Core/Repositories/RepositoryData.cs(158,35): warning SYSLIB0050: 'StreamingContext.StreamingContext(StreamingContextStates, object?)' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
/media/sf_CKAN/Core/Registry/RegistryManager.cs(316,48): warning SYSLIB0050: 'StreamingContextStates' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
/media/sf_CKAN/Core/Registry/RegistryManager.cs(316,27): warning SYSLIB0050: 'StreamingContext.StreamingContext(StreamingContextStates, object?)' is obsolete: 'Formatter-based serialization is obsolete and should not be used.' (https://aka.ms/dotnet-warnings/SYSLIB0050) [/media/sf_CKAN/Core/CKAN-core.csproj::TargetFramework=net8.0]
    4 Warning(s)
    0 Error(s)

Time Elapsed 00:01:35.28

An error occurred when executing task 'Build'.

----------------------------------------
TearDown
----------------------------------------

Error: Could not resolve MSBuild.
```

doinkythederp commented:
> Ubuntu 24.04 says:
> 
> > `media/sf_CKAN/build/BuildContext.cs(128,14): error CS0121: The call is ambiguous between the following methods or properties: 'string.Split(char[]?, StringSplitOptions)' and 'string.Split(string?, StringSplitOptions)' [/media/sf_CKAN/build/Build.csproj]`
> 
> Possibly a dotnet versioning thing? This VM has 8.0.112.

Resolved in [`9cf2633` (#4320)](https://github.com/KSP-CKAN/CKAN/pull/4320/commits/9cf263343de9ef72cb17839feff0c54cf8d72326)

HebaruSan commented:
OK, I can now build on all systems, so I'm reviewing the actual changes.

Looks like the target defaulting is broken:

```
> .\build.ps1 test

Error: Unknown command 'test'.

       test
       ^^^^ No such command
```

That's one of the things that was deleted in `build` and `build.ps1`, looks like it wasn't carried forward to the new stuff?

https://github.com/KSP-CKAN/CKAN/blob/27c6d9adc9cd445f6c1f9ba5d6d551e27f6bc647/build.ps1#L46-L52

https://github.com/KSP-CKAN/CKAN/blob/27c6d9adc9cd445f6c1f9ba5d6d551e27f6bc647/build#L48-L57

doinkythederp commented:
> OK, I can now build on all systems, so I'm reviewing the actual changes.
> 
> Looks like the target defaulting is broken:
> 
> ```
> > .\build.ps1 test
> 
> Error: Unknown command 'test'.
> 
>        test
>        ^^^^ No such command
> ```
> 
>  ```
> $ ./build.sh test
> 
> Error: Unknown command 'test'.
> 
>        test
>        ^^^^ No such command
> ```
> 
> That's one of the things that was deleted in `build` and `build.ps1`, looks like it wasn't carried forward to the new stuff?
> 
> https://github.com/KSP-CKAN/CKAN/blob/27c6d9adc9cd445f6c1f9ba5d6d551e27f6bc647/build.ps1#L46-L52
> 
> https://github.com/KSP-CKAN/CKAN/blob/27c6d9adc9cd445f6c1f9ba5d6d551e27f6bc647/build#L48-L57

Oops! I think when I applied the template for Cake Frosting it must have overwritten that.

HebaruSan commented on .gitignore:
```diff
@@ -52,3 +52,111 @@ bld/
+.idea/**/uiDesigner.xml
+.idea/**/dbnavigator.xml
+
+# Gradle
+.idea/**/gradle.xml
+.idea/**/libraries
+
+# Gradle and Maven with auto-import
+# When using Gradle or Maven with auto-import, you should exclude module files,
+# since they will be recreated, and may cause churn.  Uncomment if using
+# auto-import.
+# .idea/artifacts
+# .idea/compiler.xml
+# .idea/jarRepositories.xml
+# .idea/modules.xml
+# .idea/*.iml
+# .idea/modules
+# *.iml
+# *.ipr
+
+# CMake
+cmake-build-*/
```
Can we remove these lines since we're not using Gradle or CMake?

HebaruSan commented on .gitignore:
```diff
@@ -52,3 +52,111 @@ bld/
+# SonarLint plugin
+.idea/sonarlint/
+
+# Crashlytics plugin (for Android Studio and IntelliJ)
+com_crashlytics_export_strings.xml
+crashlytics.properties
+crashlytics-build.properties
+fabric.properties
```
Android Studio and IntelliJ also sound non-applicable to this project.

HebaruSan reviewed (approved):
Thanks so much for working on this! 🚀
